### ISSUE TITLE
How do i provide a verifier with updated or corrected information previously sent in a prior presentation?

### ISSUE THREAD

**Issue by BenjaminMoe** [w3c-ccg/traceability-interop#485]
1. I sent a credential that i want to disassociate with that presentation instance
2. I sent a credential that i want to replace with another credential

**Suggestion:** 
- We use a compound key for workflow instance + vc.id
- because vc.id is unique, we need to define `{ replaces: 'previous credential id' }` to satisfy the compound key

**Requirements:**
- Needs to account for credentials that can't be re-issued
- Needs to account that previous credentials might be the correct credential
- Needs to account for credentials that might be valid in another context
- Needs to account for credentials that might be included in other workflow presentations

**Example 1:** I sent my driver's license when i meant to send my CTPAT license. 

This is a pretty stupid example, but it's illustrative. My driver's license was issued by some state, I do not have the ability to re-issue or edit either one of these credentials. The way we would address this would be to send a new presentation, with a new presentation.id, the same presentation.instance, and the same presentation.workflow.

The presentation would contain two credentials. One credential to say, "please ignore my driver's license in this workflow". And then provide the CTPAT credential. So effectively what we need is a "remove credential from instance" schema defined in trace-vocab. And then I we need somewhere to define difference cases around workflow instances and how different cases should be handled in the respec document.

**Comment by nissimsan** [comment]
I don't think these requirements correctly capture what a presentation is. Imagine the physical equivalent: I'm at the physical border, and is being asked to present my CTPAT certificate. I show my driver's license by mistake. Realizing my mistake, I pull out my CTPAT certificate and present it. 

I can't make you "unsee" what I put in front of you. But it doesn't matter, it's just a bit of extra, irrelevant data. Complex communication protocols attempting to let the holder change the verifiers data graph is dangerous path to be heading down. The verifier can easily make that judgement, ignoring irrelevant credential types. 

IMO we have the tools we need to overcome these situations: 
1. Just let the erroneous credential sit and present the required credential. 
2. Abandon the erroneous workflow instance and start a new one.

**Comment by BenjaminMoe** [comment]
> Just let the erroneous credential sit and present the required credential.

Let's throw in another example. I send the wrong invoice, and then follow up by sending the correct invoice. We have two invoices, with two different vc.id's with different content associated with the same workflow. How is a verifier supposed to know which one needs to be ignored?

> Abandon the erroneous workflow instance and start a new one.

This is another assumption that the person submitting the documents is the one controlling the workflow. The whole point of workflows is that several parties can make claims about the same workflow instance. We have one party send an intent to import, another party follows up with a mill test report, and then the third party follows up with a wrong invoice. And instead of that third party being able to correct a single credential, everyone has to start over?

And then you're also leaving the verifier on the hook not knowing that instance was abandoned. We'd need a way to say terminate presentation. Which is just compounding the same problem of "please ignore this one invoice".

**Comment by nissimsan** [comment]
The two invoices is a good example. 

I would suggest that we implement these behaviors on the Presentation-level instead of Credential-level. So, a presentation should be interpreted as a replacement of a previous presentation. 

An empty presentation would then fully cancel a previous presentation. 

Presentations are what bind Credentials to Workflows - it seems like the right place to introduce this kind of feature. It might require some minor changes: support for empty presentation and presentation id, etc. But nothing we're not in control over with TraceablePresentation. 

In the paper analogy, it means "I showed you my driver's license, I actually meant to show you my CTPAT certificate". Although you did already see the former, I'm telling you it shouldn't be considered.

**Comment by nissimsan** [comment]
`presentationId` and `workflowInstanceId` is not the same. Let me explain myself a little better: 

First: 
```
{
 presentationId: 1001
 workflowId: abcd-1234
 credentials: [cred1, cred2]
}
```

Then: 
```
{
 presentationId: 1002
 replacesPresentation: 1001
 workflowId: abcd-1234
  credentials: [cred1, cred3]
}
```

**Comment by BenjaminMoe** [comment]
Aaaah, okay. So rather than managing things on a credential by credential basis. You could resend the entire presentation, and say "discard the last presentation i sent, please use these credentials instead". This would also allow you to delete parts of a workflow by sending an empty presentation. 

I agree with this approach. I think the concrete we would need to pour around this would be the policy that we expect the presentation to be sent by the same party.

**Comment by OR13** [comment]
`replacesPresentation` seems like a good solution, it would delete 100% of the credentials from the previous presentation and insert all the new ones.

You would still have a potential to form compound keys with the other identifiers.

**Comment by brownoxford** [comment]
I'm in favor of the way this conversation is moving, although I think we should be careful about saying that a new presentation will "delete" old credentials from the previous presentation. It seems like "invalidate" is a better term and we leave it up to the system receiving the data to decide whether or not they are going to retain the old VCs?

Should we also talk about including some normative language saying that a system "should" preserve information about presentations that have been presented (otherwise how else are you going to re-send credentials in the event that you need to replace one of them).

**Comment by OR13** [comment]
> Should we also talk about including some normative language saying that a system "should" preserve information...

IMO no... this is already a bit of an overfit conversation.

`replacesPresentation` is a full document replace operation, while the previous presentation might be retained for record keeping purposes, none of its credentials or contained claims would be surfaced regarding a workflow instance moving forward, unless as part of some forensic investigation.... I don't really want to be saying anything like that.

@brownoxford @mprorock sync, and propose language... try and find a nice boundary.

**Comment by mprorock** [comment]
deferring to @brownoxford on this since he is on the hook for implementation

**Comment by mprorock** [comment]
personal take though - let's not overcomplicate things, and avoid specifying anything past the API layer and basic business rules for data processing as would be required for interop - i think things like `replacesPresentation` vs just using the same ID is way over the top and will lead to significant problems

**Comment by OR13** [comment]
@mprorock can you rephrase as informative text for respec?

**Comment by mprorock** [comment]
@OR13 yes - let's mark this ready for PR and i can work up some language

**Comment by BenjaminMoe** [comment]
>just using the same id for a credential would imply that when you are trying to assemble the latest state, you would get credential by type, etc, but key off issuance date for the credential

@mprorock @OR13 I don't think that approach works. 

1. It assumes you are the issuer. 
2. It conflicts with the decision made here that vc.id MUST be unique: https://github.com/w3c-ccg/traceability-interop/issues/468

**Comment by OR13** [comment]
we could say "presentation ids are always random, unless you are intending to overwrite".

**Comment by mprorock** [comment]
> * It assumes you are the issuer.

in a general system the issuer might change - in supply chains, especially with regulatory concerns, can you show me a case where a credential would be "updated" by a different issuer

> * It conflicts with the decision made here that vc.id MUST be unique: [Method for referencing stored verifiable credentials #468](https://github.com/w3c-ccg/traceability-interop/issues/468)

text will need to be clarified, however, if does not conflict - it is an ID per the [vc data model ](https://www.w3.org/TR/vc-data-model/#identifiers) that is unique to the issuer, subject, type, etc combination

by "replaying" the `credential.id` you are signaling that this data might be out of date or otherwise incorrect, e.g. user error, etc.  this is much simpler than trying to tell a user or someone receiving the data to perform some complex query to see if this is the latest information or not, and how to go find it to see if it is potentially an update to existing data.

there is a reason the notion of primary keys and foreign keys exist in the bulk of databases, without a way of uniquely identifying a record we are inviting a nightmare of use case specific business rules.

**Comment by mprorock** [comment]
i am writing a PR now to review

**Comment by OR13** [comment]
Ok, same deal with. credential.id... which we already have to reuse to support revocation.

**Comment by BenjaminMoe** [comment]
> by "replaying" the credential.id you are signaling that this data might be out of date or otherwise incorrect, e.g. user error, etc. this is much simpler than trying to tell a user or someone receiving the data to perform some complex query to see if this is the latest information or not, and how to go find it to see if it is potentially an update to existing data.

This does not align with the flow of the conversation or the requirements as i understand it.

This assumes I am the issuer. I might have sent an invoice from another party, and gotten a new invoice with a different vc.id. In this case I would not be able to re-issue the credential.

The second part is the suggestion and flow of this issue was to re-use the presentation.id, which is something someone would have control over.

**Comment by brownoxford** [comment]
> Ok, same deal with `credential.id`... which we already have to reuse to support revocation.

@OR13 It sounds like you are saying that we need to allow non-unique credential ID values at `/credentials/issue` endpoints, in which case we will need to modify things as currently stated in proposal 6 of #468.

**Comment by brownoxford** [comment]
@OR13 I'm still working on getting up to speed on revocation, but I agree that it makes sense to leverage existing features. We just need to modify our decision in #468 to redefine requirements around credential.id to allow it to be non-unique, and to define behavior when it is replayed.

**Comment by nissimsan** [comment]
> i think things like replacesPresentation vs just using the same ID is way over the top and will lead to significant problems

@mprorock, I'd say you are overloading `presentation.id` here. Arguments for `replacesPresentation`:
- It takes a lot of normative protocol to explain how two identical presentation.ids should be interpreted. `replacesPresentation` is explicit and self-explanatory. 
- The verifiers RDF graph won't get messed up by duplicate ids. The verifier can implement the the store and the delete operations as separate basic transactions.  
- Basing intent on sequencing is actually impossible now, oauth presentations don't have timestamps. Even assuming they did (and we do need to introduce them!) it's brittle compared to explicit referencing.

**Comment by nissimsan** [comment]
The conversation had turned back to credentials, credential.id, revocation. It seems like a separate discussion snuck into this issue. 

Re-stating the simplicity of implementing the CRUD-requirements at the presentation-level: the entire previous presentation should be dismissed, replaced by this new one. 

Even heathen id-less credentials can be replaced this way!  :)

**Comment by mkhraisha** [comment]
> > Should we also talk about including some normative language saying that a system "should" preserve information...
> 
> IMO no... this is already a bit of an overfit conversation... informative seems ok though.
> 
> `replacesPresentation` is a full document replace operation, while the previous presentation might be retained for record keeping purposes, none of its credentials or contained claims would be surfaced regarding a workflow instance moving forward, unless as part of some forensic investigation.... I don't really want to be saying anything like that.

We should absolutely NOT be deleting/replacing previous credentials that were submitted as part of an incorrect presentation, we should be disassociating the credentials from the workflow. I think that's you're trying to say which I agree with.

**Comment by BenjaminMoe** [comment]
1. MUST account for issuer and holder
1. MUST account for workflows being multiparty
1. MUST account that the credential might be valid in another context

Made some doodles to capture some possible real-world scenarios. Agree with @mkhraisha's point around disassociating the credentials from the workflow.  

![Examples - Frame 1](https://user-images.githubusercontent.com/86194145/213001352-b8bd9588-685b-4ede-b664-769f7a238f53.jpg)

![Examples - Frame 2](https://user-images.githubusercontent.com/86194145/213001384-5b181122-20c1-4b70-bb3f-6f1d5fbedd8d.jpg)

![Examples - Frame 3](https://user-images.githubusercontent.com/86194145/213001404-7f156121-2aad-411a-8ace-a598e1ba2ba4.jpg)

**Comment by mprorock** [comment]
> > i think things like replacesPresentation vs just using the same ID is way over the top and will lead to significant problems
> 
> @mprorock, I'd say you are overloading `presentation.id` here. Arguments for `replacesPresentation`:
> 
> * It takes a lot of normative protocol to explain how two identical presentation.ids should be interpreted. `replacesPresentation` is explicit and self-explanatory.
> * The verifiers RDF graph won't get messed up by duplicate ids. The verifier can implement the the store and the delete operations as separate basic transactions.
> * Basing intent on sequencing is actually impossible now, oauth presentations don't have timestamps. Even assuming they did (and we do need to introduce them!) it's brittle compared to explicit referencing.

i think you misinterpret - `presentation.id` should always be unique - think of it as a database transaction id

`credential.id`, that is different, and is totally fine to have dupes of (as is the case with revocation / status)

**Comment by OR13** [comment]
```json
{
  "@context": [
    "https://www.w3.org/2018/credentials/v1",
    "https://w3id.org/traceability/v1"
  ],
  "id": "urn:uuid:presentation-123",
  "type": [
    "VerifiablePresentation",
    "TraceablePresentation"
  ],
  "workflow": {
    "definition": [
      "urn:uuid:definition-123"
    ],
    "instance": [
      "urn:uuid:instance-123"
    ],
  },
  "holder": {
    "id": "did:web:sender.example"
  },
  "verifiableCredential": [
    {
      "@context": [
        "https://www.w3.org/2018/credentials/v1",
        "https://w3id.org/traceability/v1"
      ],
      "id": "urn:uuid:credential-123",
      "type": [
        "VerifiableCredential",
        "MillTestReportCredential"
      ],
      "issuer": {
        "id": "did:web:manufacturer.example",
      },
      "issuanceDate": "2022-06-06T08:10:00+00:00"
    }
  ]
}
```

Need to figure out how to make events of the format "last write wins".

**Comment by mprorock** [comment]
> Need to figure out how to make events of the format "last write wins".

+1 - clarifying "last" as determined by `verifiableCredential.issuanceDate`

**Comment by BenjaminMoe** [comment]
> this is a different problem than how do i supersede a VC that has been previously presented with an update or change to information

This was the intention of the issue. I will update the description to match this.

**Comment by brownoxford** [comment]
Discussed on call - would like to get more participation before we continue to tackle this issue.

**Comment by nissimsan** [comment]
All you can do is: 
- Send the right data
- Notify out of band, please consider this a replace
... You can't expect the verifier to actually delete; only aid in how to interpret.

**Comment by BenjaminMoe** [comment]
Analogy: in health care, sometimes you send the wrong patient records. Out of band, you need to communicate to the provider to do the right thing with the data that was sent in error.

**Comment by OR13** [comment]
![Screen Shot 2023-08-22 at 1 06 01 PM](https://github.com/w3c-ccg/traceability-interop/assets/8295856/f4c59218-cc4f-4e3a-b21c-488d29ce29a2)

**Comment by OR13** [comment]
issue marker needs to be removed, JSON-LD graph needs to be reviewed, and confirmed to be correct.

**Comment by nissimsan** [comment]
![image](https://github.com/w3c-ccg/traceability-interop/assets/34443212/73a3b8dd-c44e-473e-9ef4-2d0fe65f1527)
We have solved the replace ^

**Comment by OR13** [comment]
We have the examples here, which are demonstrating the recommended approach. https://github.com/transmute-industries/transmute/tree/main/examples/neo4j#traceable-presentations
